File: python/algorithm/DisjointSet.py


class DisjointSet :

	def __init__ (self, n) :

		self.__tmp = []
		for i in range(n) :
			self.__tmp.append({i})

		return

	def find(self, a) :
		for s in self.__tmp :
			if a in s :
				return s

		return None;

	def findIdx(self, a) :
		for i in range(len(self.__tmp)) :
			if a in (self.__tmp)[i] :
				return i
		return -1

	def merge(self, a, b) :

		sA = self.find(a)
		sB = self.find(b)

		iA = self.findIdx(a)
		iB = self.findIdx(b)

		(self.__tmp)[iB] |= sA

		del (self.__tmp)[iA]

		return



if __name__ == "__main__" :

	tmp = DisjointSet(10)
	tmp.merge(0,1)
	print(tmp.find(0))

File: c++/algorithm/DisjointSet.cpp
#include <iostream>
#include <vector>

using namespace std;

template <typename T>
struct Node {
  Node* representitive;
  T value;
  Node* next;

  Node(T value) : representitive(this), value(value), next(nullptr) {}
};

template <typename T>
struct DisjointSet {
  DisjointSet(int n) {}

  virtual int find(int a) = 0;
  virtual bool merge(int a, int b) = 0;
};

template <typename T>
struct LinkedListDisjointSet : DisjointSet<T> {
  vector<Node<T>*> groups;

  LinkedListDisjointSet(int n) : DisjointSet<T>(n), groups(0) {
    for (int i = 0; i < n; i++) {
      groups.push_back(new Node<T>(i));
    }
  }

  void _changeRoot(int rootA, int rootB) {
    if (find(rootA) != rootA || find(rootB) != rootB) return;

    Node<T>* cur = groups[rootA];
    while (cur->next != nullptr) {
      cur = cur->next;
    }
    cur->next = groups[rootB];
    cur = groups[rootB];

    while (cur != nullptr) {
      cur->representitive = groups[rootA];
      cur = cur->next;
    }
  }

  int find(int a) override {
    Node<T>* r = groups[a]->representitive;
    int i = 0;
    for (auto group : groups) {
      if (r == group) break;
      i++;
    }

    return i;
  }

  bool merge(int a, int b) override {
    int rootA = find(a);
    int rootB = find(b);

    if (rootA == rootB) return false;

    if (a < b)
      _changeRoot(rootA, rootB);
    else
      _changeRoot(rootB, rootA);

    return true;
  }
};

int main(void) {
  LinkedListDisjointSet<int> llds(10);

  for (int i = 0; i < 10; i++) {
    cout << llds.find(i) << " ";
  }
  cout << endl;

  llds.merge(0, 1);
  llds.merge(1, 2);
  llds.merge(8, 9);

  for (int i = 0; i < 10; i++) {
    cout << llds.find(i) << " ";
  }
  cout << endl;

  return 0;
}